wordpress-mobile/WordPress-Android
# My Site Dashboard: Phase 2: Metrics - Track Cards Shown

Fixes #15732 

This PR implements shown tracking as follows:
- **MY_SITE_DASHBOARD_CARD_SHOWN** event for dashboard "post" cards with the following properties
     - type = “post”
     - sub_type = “create_first”, “create_next”, “drafts”, “scheduled”

- Tracking for Dashboard ErrorCard
**MY_SITE_DASHBOARD_CARD_SHOWN** event with the following properties
     - type = “error”

- Moved the **DOMAIN_CREDIT_PROMPT_SHOWN** event from `CardsBuilder` to `CardsShownTracker`.
- Added helper class `CardsShownTracker`  
- Tracking whether or not a shown card event was tracked is encapsulated in `CardsTracker` (for dashboard cards) and in `CardsShownTracker` (for non-dashboard cards).

**MERGE INSTRUCTIONS**
I put the **Not Ready For Merge** label on this to leave room for any possible direction changes. If there are no direction changes, then please remove the label and treat it as a normal PR. Thanks. 🙇 


**Notes**:
- Cards shown are tracked 1x for each site
- Cards shown events are not tracked onResume or return from background (unless the OS kills the process)
- Cards shown events are cleared when the selected site changes
- If a user is shown the create first post card, then creates a post, we will track create next after it is shown. Same goes for drafts or scheduled. Each card can trigger a shown 1x per site change. 
- Tapping the site selector and not changing the site, will not trigger a new tracking.
- The domain registration card is shown tracking event will follow the same pattern as the dashboard cards. Previously this event was tracked during onResume. We should note this in the tracks posts. 

**Other implementation Notes**
- I was going back-n-forth on naming the `CardsShownTracker`. I was going to name it `CardsTracker` and move other card tracking events to it, but there wasn't anything to add Also I didn't like the name conflict with `CardsTracker` under the dashboard package. As always, I am open to changing names.
- I thought about changing the way "domain" shown is tracked; using a **MY_SITE_CARD_SHOWN**  pattern that mimics the **MY_SITE_DASHBOARD_CARD_SHOWN**; however I didn't want to throw off any funnels that depend on it. We could "double" track that with the old event name + implement a new way. Would love thoughts on that?
- I went with this implementation because we are always recreating the UiModel/State fresh (no 'copy") . I couldn't use partial state, because that occurs before the card is built. If you can think of another implementation, I am more than open to switching. Tried to keep it simple and to the point.
- I thought about using **MY_SITE_DASHBOARD_ERROR_CARD_SHOWN**, but went with the type property instead. In the future, when we are tracking individual "card" errors, we can easily introduce a subtype. 
- I did not track Posts as `PostCardWithItems` or `PostCardWithoutItems`, I felt that "post" along with subtype was the better option (it's self explanatory)
 
I know this was a lot of chatter, but thanks for making it to the bottom. This 🤗  is for you!!

**To test:**
Setup
- Launch the app
- Navigate to Me -> App Settings -> Debug Settings
- Ensure that the `MySiteDashboardPhase2FeatureConfig` is to set to on
- Restart the app if needed
- Navigate to Me -> App Settings -> Privacy Settings 
- Toggle the "collect information" switch to on
------------------------
- Navigate to the My Site tab
- Use the Site selector to chose a site that has no posts
- Navigate to Me -> Help and Support -> Application log
- Note that the following is shown in the log
   I: 🔵 Tracked: my_site_dashboard_card_shown, Properties: {"type":"post","subtype":"create_first"}
-----------------------
- Navigate to the My Site tab
- Use the Site selector to chose a site has posts, but no draft or scheduled posts
- Navigate to Me -> Help and Support -> Application log
- Note that the following is shown in the log
I: 🔵 Tracked: my_site_dashboard_card_shown, Properties: {"type":"post","subtype":"create_next"}
-----------------------
- Navigate to the My Site tab
- Use the Site selector to chose a site that has draft posts
- Navigate to Me -> Help and Support -> Application log
- Note that the following is shown in the log
I: 🔵 Tracked: my_site_dashboard_card_shown, Properties: {"type":"post","subtype":"draft"}
-----------------------
- Navigate to the My Site tab
- Use the Site selector to chose a site that has scheduled posts
- Navigate to Me -> Help and Support -> Application log
- Note that the following is shown in the log
I: 🔵 Tracked: my_site_dashboard_card_shown, Properties: {"type":"post","subtype":"scheduled"}
---------------------

I couldn't trigger the error card in the app, so if to test you need to pull down this PR and adjust CardsBuilder.build to only create the `ErrorCard`. Run the app, navigate to My Site and then check the application log for :
I: 🔵 Tracked: my_site_dashboard_card_shown, Properties: {"type":"error"}

------------------
- Using any of the testing scenarios above (excluding error, unless you feel like setting up the logic)
- Note the application log has tracked card shown
- Turn your device to trigger a configuration change
- Navigate to application log
- Note the event didn't trigger again 

## Regression Notes
1. Potential unintended areas of impact
Cards shown are not properly tracked

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual and automated tests

3. What automated tests I added (or what prevented me from doing so)
Add tests to MySiteViewModel, CardsTracker, CardsShownTracker.

PR submission checklist:
